BITSWAP.md

```diff
@@ -1,200 +1,200 @@
-  repeated Block payload = 3; // used to send Blocks in bitswap 1.1.0
-}
+  repeated bytes blocks = 2;
 ```
 
-## Want-Manager
-
-The want-manager handles requests for blocks. For a requested block, identified by `cid`, the `Bitswap.GetBlock(cid)` method is invoked. `Bitswap.GetBlock(cid)` requests `cid` from the network and, if the corresponding `Block` is received, returns it. More concretely, `Bitswap.GetBlock(cid)` adds `cid` to our wantlist. The want-manager then updates all peers with this addition by adding a new `Entry` to each peer’s message queue, who then may or may not respond with the desired block.
-
-## Decision Engine
-
-The decision engine decides how to allocate resources to peers. When a `Message` with a wantlist from a peer is received, the `Message` is sent to the decision engine. For every `CID` in the wantlist that we have the corresponding `Block` for, the block has a `Task` added to the peer’s `TaskQueue`. A `Task` is considered complete once the corresponding `Block` has been sent to the message queue for that peer.
-
-The primary data structure in the decision engine is the peer request queue (`PRQ`). The `PRQ` adds peers to a weighted round-robin queue, where the weights are based on one or more peer-specific metrics. This is where Bitswap *strategies* come in. Currently, a `Strategy` is a function whose input is a peer’s `Ledger` and output is a weight for that peer. The peers are then served the `Task`s in their respective `TaskQueue`s. The amount of data each peer is served in a given round-robin round is determined by their relative weight in the queue. The in-progress Bitswap strategy implementation can be found [here](https://github.com/ipfs/research-bitswap/tree/docs/strategy_impl/strategy-impl). Further Bitswap strategy metrics and configuration interfaces are planned for the near future.
+### Protocol Format
 
-*Note: The Bitswap strategy implementations in the current releases of* `go-ipfs` *and* `js-ipfs` *do not conform to the description here as of the time of this writing.*
+All protocol messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
 
-## Message Queue
+All protocol messages must be less than or equal to 4MiB in size
 
-Each active peer has an associated message queue. The message queue holds the next message to be sent to that peer. The message queues receive updates from two other subsystems:
+### Block Sizes
 
-1.  Wantlist manager: When a `CID` is added or removed from our wantlist, we must update our peers – these wantlist updates are sent to all relevant peers’ message queues.
-2.  Decision engine: When we have a block that a peer wants and the decision engine decides to send the block, we propagate the block to that peer’s message queue.
+Bitswap implementations must support sending and receiving individual blocks of sizes up to 2MiB
```

Comment:
I have already seen other IPLD implementers saying that the limit is actually 2MiB-1byte, (in order to fit in a 21bits uint ?).

However I couldn't find any data to back this up, current implementations accepts more anyway to account for the fact that very few chunkers are capable to precisely matching this, and might be slightly over that limit.

It should be clearer if that is `2MiB` inclusive or exclusive.

Reply:
Added more clarity in https://github.com/ipfs/specs/pull/269/commits/c9b1451cdfbfc937424edb520518ec0c2e17b1fe